File: move.py
from __future__ import annotations
from constants import PieceType
from typing import TYPE_CHECKING, Optional
from square import Square
from utils import pieceToLetter

if TYPE_CHECKING:
    from piece import Piece

class Move:
    def __init__(self, fromSqr: Square, toSqr: Square, newPiece: Optional[PieceType] = None, capturedPiece: Optional[Piece] = None):
        assert fromSqr.piece is not None, f"Invalid Move: there is no piece on the source square {fromSqr}"
        self.piece: Piece = fromSqr.piece
        self.fromSqr = fromSqr
        self.toSqr = toSqr
        self.newPiece = newPiece
        self.isPromotion = self.piece.kind == PieceType.PAWN and (self.toSqr.rowIdx == 0 or self.toSqr.rowIdx == 7)
        self.isCastling = self.piece.kind == PieceType.KING and abs(self.fromSqr.colIdx - self.toSqr.colIdx) > 1
        self.isEnPassant = self.piece.kind == PieceType.PAWN and self.toSqr.piece is None and self.fromSqr.colIdx != self.toSqr.colIdx
        if not self.isEnPassant:
            self.capturedPiece = self.toSqr.piece
        else:
            self.capturedPiece = capturedPiece

    def __str__(self) -> str:
        return f'{str(self.piece)}{self.fromSqr}-{self.toSqr}{pieceToLetter[self.newPiece].upper() if self.newPiece else ""}'

    def __repr__(self) -> str:
        return f'Move({self.fromSqr}, {self.toSqr}, {self.newPiece}, {self.capturedPiece})'

File: c_engine/src/move.cpp
#include "move.h"
#include "piece.h"
#include "square.h"
#include "utils.h"


Move::Move(Piece *piece, Square *toSqr, std::optional<PieceType> newPiece)
: _piece(piece),
  _toSqr(toSqr),
  _newPiece(newPiece) {

    ASSERT(piece->get_square() != nullptr, "piece has no square");
    _fromSqr = piece->get_square();

    // if the piece is a pawn and it's being promoted, the newPiece must be specified
    ASSERT(newPiece.has_value() || (piece->_kind != PieceType::PAWN || toSqr->get_coordinate().row != (piece->_color == Color::WHITE ? 7 : 0)),
        "a move must specify the new piece iff it's a promotion");
}

bool Move::operator==(Move &other) const {
    return _piece == other._piece && _fromSqr == other._fromSqr && _toSqr == other._toSqr && _newPiece == other._newPiece;
}

bool Move::is_promotion() const {
    return _newPiece.has_value();
}

bool Move::is_castling() const {
    return _piece->_kind == PieceType::KING && abs(_fromSqr->get_coordinate().col - _toSqr->get_coordinate().col) == 2;
}

std::string Move::str() const {
    return _piece->str().substr(0, 1) + _fromSqr->str() + "-" + _toSqr->str() + (is_promotion() ? piece_type_to_letter(_newPiece.value()) : "");
}

std::ostream& operator << (std::ostream &os, const Move &move) {
    return os << move.str();
}

bool operator==(Move &lhs, Move &rhs) {
    return lhs._piece == rhs._piece && lhs._fromSqr == rhs._fromSqr && lhs._toSqr == rhs._toSqr && lhs._newPiece == rhs._newPiece;
}
